Data/models/Bookmark.swift

```diff
@@ -96,59 +96,62 @@ public final class Bookmark: NSManagedObject, WebsitePresentable, Syncable, CRUD
     // Syncable
     public func update(syncRecord record: SyncRecord?) {
         guard let bookmark = record as? SyncBookmark, let site = bookmark.site else { return }
-        title = site.title
-        update(customTitle: site.customTitle, url: site.location, newSyncOrder: bookmark.syncOrder)
-        lastVisited = Date(timeIntervalSince1970: (Double(site.lastAccessedTime ?? 0) / 1000.0))
-        syncParentUUID = bookmark.parentFolderObjectId
-        // No auto-save, must be handled by caller if desired
+        
+        DataController.performBackgroundTask { context in
+            
+            guard let bookmarkToUpdate = context.object(with: self.objectID) as? Bookmark else { return }
+            
+            bookmarkToUpdate.title = site.title
+            bookmarkToUpdate.update(customTitle: site.customTitle, url: site.location,
+                                    newSyncOrder: bookmark.syncOrder, context: context)
+            bookmarkToUpdate.lastVisited = Date(timeIntervalSince1970: (Double(site.lastAccessedTime ?? 0) / 1000.0))
+            bookmarkToUpdate.syncParentUUID = bookmark.parentFolderObjectId
+        }
     }
     
     public func update(customTitle: String?, url: String?, newSyncOrder: String? = nil, save: Bool = false,
-                       sendToSync: Bool = false) {
-        var contextToUpdate: NSManagedObjectContext?
-        
-        // Syncable.update() doesn't save to CD at the moment, we need to use managedObjectContext here.
-        if save == false {
-            contextToUpdate = managedObjectContext
-            // Updated object usually uses view context, all database writes should happen on a background thread
-            // so we need to fetch the object using background context.
-        } else if managedObjectContext?.concurrencyType != .privateQueueConcurrencyType {
-            contextToUpdate = DataController.newBackgroundContext()
-        }
+                       sendToSync: Bool = false, context: NSManagedObjectContext? = nil) {
         
-        guard let bookmarkToUpdate = (try? contextToUpdate?.existingObject(with: objectID)) as? Bookmark,
-            let context = contextToUpdate else {
-                return
+        if let context = context {
+            updateImpl(customTitle: customTitle, url: url, newSyncOrder: newSyncOrder, save: save, sendToSync: sendToSync, context: context)
+        } else {
+            DataController.performBackgroundTask { [weak self] backgroundContext in
```

Comment:
Don't think we should use a `weak self` here, if something else deallocated this bookmark from memory it would never get updated in CD.